importccl: add more errors to transient retryable error list

adityamaru opened the pull request:
This change adds a few more errors that were exposed when stressing
a previously skipped test, that can occur when a worker node running
the import is shutdown. These errors should not cause the job to fail
but instead trigger a replan.

I ran 500+ iterations under stress of TestImportWorkerFailure.

Release note: None

adityamaru committed:
importccl: add more errors to transient retryable error list

This change adds a few more errors that were exposed when stressing
a previously skipped test, that can occur when a worker node running
the import is shutdown. These errors should not cause the job to fail
but instead trigger a replan.

I ran 500+ iterations under stress of TestImportWorkerFailure.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64136)
<!-- Reviewable:end -->

adityamaru commented:
I want backup + restore to use this list as well, and have a similar unit test, but I'm seeing leaked goroutines when writing a similar test for backup. I think it is because of the incorrect teardown of the pre-retry flow. Still tracking down what can be done about that/

adityamaru committed:
backupccl: add test that shuts down worker during BACKUP

This change adds a unit test that shuts down a worker to test
that the backup job retries when it sees certain retryable
errors.

This test uncovered a situation in which a backup data processor
could be moved to state other than StatusRunning, that caused it
to exit early from the Next() method. The goroutines responsible
for exporting the spans get stuck attempting to push to the progCh
that is only drained after the above StatusRunning check. This
change drains the progCh at the beginning of the Next() method.

Release note: None

miretskiy commented on pkg/sql/flowinfra/flow_registry.go:
```diff
@@ -28,7 +28,15 @@ import (
+// ErrNoInboundStreamConnection.
+func IsNoInboundStreamConnectionError(err error) bool {
+	return errors.Is(err, ErrNoInboundStreamConnection)
+}
```
It's strange to have both the error exported as well as a IsError function to test.
I think it's one or the other: either keep the ErrNoInboundStreamConnection unexported (and keep Is method), or just export Err...

adityamaru replied:
Done.

pbardea commented on pkg/ccl/backupccl/backup_processor.go:
```diff
@@ -118,17 +118,17 @@ func (bp *backupDataProcessor) Start(ctx context.Context) {
 		return nil, &execinfrapb.ProducerMetadata{BulkProcessorProgress: &p}
 	}
 
+	if bp.State != execinfra.StateRunning {
```
I'm curious what motivated moving this down?

adityamaru replied:
@pbardea added an explanation to the commit message:
```
This test uncovered a situation in which a backup data processor
could be moved to state other than StatusRunning, which caused it
to exit early from the Next() method without draining the progCh. The goroutines responsible
for exporting the spans get stuck attempting to push to the progCh
that is only drained after the above StatusRunning check. This
change drains the progCh at the beginning of the Next() method.
```

Does this make sense? It seemed to mitigate the leaked goroutine in the backup test but wanted to run this by you.

pbardea replied:
Got it, nice find. I think a lot of the bulk processors use this pattern so it may also need updating elsewhere.

adityamaru commented:
@pbardea a couple of changes since the last review:
- I thought about Next() waiting on progCh and it seems wasteful for a backup/import to allow the worker goroutines to complete all their work only to throw it away (which would be the case if the processor is not in a Running state). The reason why restore/import were not running into this leaked goroutine was that the site where the goroutines pushed to progCh also had a case catching <-ctx.Done(). So eventually, the goroutines would be cleaned up once the flow was torn down. I added this case for backup too, and moved the progCh draining back to where it originally was.

- Stressing this test shook out a couple more things such as https://github.com/cockroachdb/cockroach/issues/64508 and https://github.com/cockroachdb/cockroach/pull/64136/files#diff-5141ba27e64a504e9d21394ba8b2cb84f0c5b79a6ecd5f2c02aaa18fda6fd704R486. The latter resulted in an NPE if we can't resolve a manifest on resume, because the backup() method accesses fields of a nil manifest.

pbardea reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64136)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @miretskiy)



<!-- Sent from Reviewable.io -->

pbardea commented:
> The latter resulted in an NPE if we can't resolve a manifest on resume, because the backup() method accesses fields of a nil manifest.

Nice catch. Do you happen to have logs pointing to as to why we would not be able to grab the manifest? I'm curious when this happens since another possibility to consider is that we could just use the previous version of the manifest (which we should have) when we can't reload it. Although, not being able to read the manifest does seem pretty indicative of something going wrong.

adityamaru commented:
> > The latter resulted in an NPE if we can't resolve a manifest on resume, because the backup() method accesses fields of a nil manifest.
> 
> Nice catch. Do you happen to have logs pointing to as to why we would not be able to grab the manifest? I'm curious when this happens since another possibility to consider is that we could just use the previous version of the manifest (which we should have) when we can't reload it. Although, not being able to read the manifest does seem pretty indicative of something going wrong.

I don't have the logs but it was a `checksum mismatch` from what I remember. I'll try and reproduce it.

adityamaru commented:
TFTR!

bors r=pbardea
